In [263]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [264]:
data = pd.read_csv('cia_world_factbook.csv')

In [265]:
data.head()

Country Area(sq km) Birth rate(births/1000 population)  \
0       String      double                             double   
1  Afghanistan      647500                              47.02   
2     Akrotiri         123                                NaN   
3      Albania       28748                              15.08   
4      Algeria     2381740                              17.13   

  Current account balance Death rate(deaths/1000 population) Debt - external  \
0                  double                             double          double   
1                     NaN                              20.75      8000000000   
2                     NaN                                NaN             NaN   
3              -504000000                               5.12      1410000000   
4             11900000000                                4.6     21900000000   

  Electricity - consumption(kWh) Electricity - production(kWh)      Exports  \
0                         double                        double       double   
1                      652200000                     540000000    446000000   
2                            NaN                           NaN          NaN   
3                     6760000000                    5680000000    552400000   
4                    23610000000                   25760000000  32160000000   

            GDP  ... Oil - production(bbl/day) Oil - proved reserves(bbl)  \
0        double  ...                    double                     double   
1   21500000000  ...                         0                          0   
2           NaN  ...                       NaN                        NaN   
3   17460000000  ...                      2000                  185500000   
4  212300000000  ...                   1200000                11870000000   

  Population Public debt(% of GDP) Railways(km)  \
0     double                double       double   
1   29928987                   NaN          NaN   
2        NaN                   NaN          NaN   
3    3563112                   NaN          447   
4   32531853                  37.4         3973   

  Reserves of foreign exchange & gold Telephones - main lines in use  \
0                              double                         double   
1                                 NaN                          33100   
2                                 NaN                            NaN   
3                          1206000000                         255000   
4                         43550000000                        2199600   

  Telephones - mobile cellular Total fertility rate(children born/woman)  \
0                       double                                    double   
1                        15000                                      6.75   
2                          NaN                                       NaN   
3                      1100000                                      2.04   
4                      1447310                                      1.92   

  Unemployment rate(%)  
0               double  
1                  NaN  
2                  NaN  
3                 14.8  
4                 25.4  

[5 rows x 45 columns]

In [266]:
data= data.iloc[1:] ## removed the first row that contained the data type

In [267]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 263 entries, 1 to 263
Data columns (total 45 columns):
Country                                           263 non-null object
Area(sq km)                                       263 non-null object
Birth rate(births/1000 population)                225 non-null object
Current account balance                           149 non-null object
Death rate(deaths/1000 population)                225 non-null object
Debt - external                                   201 non-null object
Electricity - consumption(kWh)                    215 non-null object
Electricity - production(kWh)                     213 non-null object
Exports                                           224 non-null object
GDP                                               230 non-null object
GDP - per capita                                  230 non-null object
GDP - real growth rate(%)                         212 non-null object
HIV/AIDS - adult prevalence rate(%)               168 n

In [268]:
data.describe() ## a lot of missing data is present

Country Area(sq km) Birth rate(births/1000 population)  \
count       263         263                                225   
unique      263         252                                218   
top     Nigeria           5                              16.83   
freq          1           5                                  3   

       Current account balance Death rate(deaths/1000 population)  \
count                      149                                225   
unique                     149                                210   
top                 -843100000                               6.26   
freq                         1                                  3   

       Debt - external Electricity - consumption(kWh)  \
count              201                            215   
unique             188                            212   
top                  0                        4650000   
freq                 5                              2   

       Electricity - production(kWh)    Exports       GDP  ...  \
count                            213        224       230  ...   
unique                           208        220       228  ...   
top                          5000000  155000000  60000000  ...   
freq                               2          2         2  ...   

       Oil - production(bbl/day) Oil - proved reserves(bbl) Population  \
count                        212                         97        238   
unique                        94                         87        237   
top                            0                          0        361   
freq                         114                          8          2   

       Public debt(% of GDP) Railways(km) Reserves of foreign exchange & gold  \
count                    108          134                                 147   
unique                   104          133                                 146   
top                     74.3          699                          1206000000   
freq                       2            2                                   2   

       Telephones - main lines in use Telephones - mobile cellular  \
count                             232                          228   
unique                            222                          213   
top                              7000                            0   
freq                                3                            8   

       Total fertility rate(children born/woman) Unemployment rate(%)  
count                                        225                  192  
unique                                       166                  124  
top                                         1.39                   15  
freq                                           6                    6  

[4 rows x 45 columns]

In [269]:
count = data['Country'].count()
print('Total number of countries in the world is:', count)

Total number of countries in the world is: 263


In [270]:
data['Population']

1       29928987
2            NaN
3        3563112
4       32531853
5          57881
6          70549
7       11190786
8          13254
9            NaN
10         68722
11      39537943
12       2982904
13         71566
14           NaN
15      20090437
16       8184691
17       7911974
18        301790
19        688345
20           NaN
21     144319628
22        279254
23           NaN
24      10300483
25      10364388
26        279457
27       7460025
28         65365
29       2232291
30       8857870
         ...    
234     36766356
235     65444371
236      5681519
237         1405
238       112422
239      1088644
240          NaN
241     10074951
242     69660559
243      4952081
244        20556
245        11636
246     27269482
247     47425336
248      2563212
249     60441457
250    295734134
251      3415920
252     26851195
253       205754
254     25375281
255     83535576
256       108708
257          NaN
258        16025
259      2385615
260       273008
261     207270

In [271]:
pop_data = data['Population']
pop_data.isnull()

1      False
2       True
3      False
4      False
5      False
6      False
7      False
8      False
9       True
10     False
11     False
12     False
13     False
14      True
15     False
16     False
17     False
18     False
19     False
20      True
21     False
22     False
23      True
24     False
25     False
26     False
27     False
28     False
29     False
30     False
       ...  
234    False
235    False
236    False
237    False
238    False
239    False
240     True
241    False
242    False
243    False
244    False
245    False
246    False
247    False
248    False
249    False
250    False
251    False
252    False
253    False
254    False
255    False
256    False
257     True
258    False
259    False
260    False
261    False
262    False
263    False
Name: Population, Length: 263, dtype: bool

In [272]:
clean_data = pop_data.dropna()
count = clean_data.count()
print(count)

238


In [273]:
clean_data.sum() ## population sum has been summed as a string rather than an integer, lets convert

'299289873563112325318535788170549111907861325468722395379432982904715662009043781846917911974301790688345144319628279254103004831036438827945774600256536522322918857870402547616401151861127942264337236174503491392531342909464637060913607069163800053280504141822444270379989798264191598091213063138123616284295427967124760085804303912621388401617317298040449590411346670780133102411385432335476703690298950034104088013363593775057566704932535881456159913328937305328645703041829674696289335452234426065617819550627048513892011593256137628946774018243139021029853278841066835456375895024487131685641465518965228946786614160277652838121622921697520468986861000683529673710802643882419738796801786026074906401567662768835810303327318321274172449081236157597321518584433829590103092229121774842264423356485146281621714122902373826018186703534822115765563337173596617468571449198204526218040341121589242395313634910612291529398534750495907143290030868591230602193633106202903108105445542132409279127293413

In [274]:
clean_data = clean_data.astype(int) ## converting population from string data type to int

In [275]:
sum = clean_data.sum()## lets calculate the average population for a country ( some countries have been emitted from the list)
print(sum)

6903176002


### Population came around 6.9 billion which is quite close to the current world's population. Even though there were around 25 omissions, the countries ommitted seem to have a very small population and considering we'll building a recommender system for a holiday destination, ideally we refer countries that constantly monitor and publish the demographics of that country. Not that demographics are super important when it comes to making a holiday choice, but it shows that the countries might have infrastructure instability or complete inexistence of infrastructure ( no infrastructure = no commercial airport = no direct route :( ). Lets see the list of countries that didnt have any population records

In [276]:
null_data = data[['Country','Population']]
null_data = null_data[null_data.isna().any(axis=1)]
print (null_data)

                                          Country Population
2                                        Akrotiri        NaN
9                                      Antarctica        NaN
14                    Ashmore and Cartier Islands        NaN
20                                   Baker Island        NaN
23                                Bassas da India        NaN
33                                  Bouvet Island        NaN
35                 British Indian Ocean Territory        NaN
52                              Clipperton Island        NaN
66                                       Dhekelia        NaN
78                                  Europa Island        NaN
87            French Southern and Antarctic Lands        NaN
95                               Glorioso Islands        NaN
107             Heard Island and McDonald Islands        NaN
111                                Howland Island        NaN
122                                     Jan Mayen        NaN
124                     

### Quite remote places, maybe its not best we include these countries in our recommendation model. Lets continue on to find the average and median population of all our other countries

In [277]:
average = sum/count ## The average population is equal to 29 million
print(average)

29004941.18487395


In [278]:
import statistics
sorted_data = clean_data.sort_values()
sorted_data

193            46
126           361
51            361
53            628
108           921
237          1405
181          1828
180          2166
227          2701
80           2967
205          7012
202          7460
166          9341
245         11636
170         13048
8           13254
258         16025
186         20303
244         20556
58          21388
36          22643
94          27884
208         28880
164         32409
143         33717
203         38958
46          44270
81          46962
97          56375
5           57881
          ...    
222      40341462
40       42909464
54       42954279
220      44344136
247      47425336
134      48422644
120      58103033
56       60085804
249      60441457
84       60656178
235      65444371
116      68017860
242      69660559
77       73053286
72       77505756
92       82431390
255      83535576
192      87857473
160     106202903
123     127417244
179     128771988
200     143420309
21      144319628
185     162419946
34      18

In [279]:
median = statistics.median(sorted_data) 
               ## median is 4.44 million
print(median)  ## we can create a function where we pass in the data frame as a parameter
               ## take the length of the data passed, divide it by 2, round it, then use iloc to print out the median data.

4440570.5


### The median in this case is much better for estimating the average size of a population. The massive populations of USA, China, India will inflate the statistics. 

In [280]:
data2 = data[['Country','Population']]
data2 = data2.dropna()
data2['Population'] = data2['Population'].astype(int) ## converting population from string data type to int
data2= data2.sort_values(by=['Population'])

In [281]:
data2.tail(10) ## 10 most populous countries

Country  Population
179          Nigeria   128771988
200           Russia   143420309
21        Bangladesh   144319628
185         Pakistan   162419946
34            Brazil   186112794
115        Indonesia   241973879
250    United States   295734134
79   #European Union   457030418
114            India  1080264388
50             China  1306313812

### European Union should not be included, ideally this should be removed and the average pop and median should be updated, but we'll leave it there for now

In [282]:
unemp_rate = data[['Country','Unemployment rate(%)']]
unemp_rate = unemp_rate.dropna()
unemp_rate['Unemployment rate(%)'] = unemp_rate['Unemployment rate(%)'].astype(float) 
unemp_rate = unemp_rate.sort_values(by='Unemployment rate(%)')
unemp_rate.tail(5)
## top 5 documented unemployment rates

Country  Unemployment rate(%)
243             Turkmenistan                  60.0
53   Cocos (Keeling) Islands                  60.0
263                 Zimbabwe                  70.0
141                  Liberia                  85.0
170                    Nauru                  90.0

In [283]:
internet_users = data[['Country','Internet users','Population']]
internet_users = internet_users.dropna()
internet_users['Population'] = internet_users['Population'].astype(int) 
internet_users['Internet users'] = internet_users['Internet users'].astype(int) 
internet_users['internet users per capita'] = internet_users['Internet users']/internet_users['Population']
internet_users = internet_users.sort_values(by='internet users per capita')
internet_users.tail(20)

Country  Internet users  Population  internet users per capita
110      Hong Kong         3212800     6898686                   0.465712
15       Australia         9472000    20090437                   0.471468
92         Germany        39000000    82431390                   0.473121
44          Canada        16110000    32805041                   0.491083
164         Monaco           16000       32409                   0.493690
208     San Marino           14300       28880                   0.495152
183         Norway         2288000     4593041                   0.498145
83         Finland         2650000     5223442                   0.507328
65         Denmark         2756000     5432335                   0.507332
173    Netherlands         8500000    16407491                   0.518056
215      Singapore         2310000     4425720                   0.521949
176    New Zealand         2110000     4035461                   0.522865
28         Bermuda           34500       65365                   0.527805
81   Faroe Islands           25000       46962                   0.532345
250  United States       159000000   295734134                   0.537645
229         Sweden         5125000     9001774                   0.569332
143  Liechtenstein           20000       33717                   0.593173
232         Taiwan        13800000    22894384                   0.602768
134    Korea South        29220000    48422644                   0.603437
113        Iceland          195000      296737                   0.657148

In [284]:
data['Internet users per capita'] = internet_users['internet users per capita']
data.head()

Country Area(sq km) Birth rate(births/1000 population)  \
1     Afghanistan      647500                              47.02   
2        Akrotiri         123                                NaN   
3         Albania       28748                              15.08   
4         Algeria     2381740                              17.13   
5  American Samoa         199                              23.13   

  Current account balance Death rate(deaths/1000 population) Debt - external  \
1                     NaN                              20.75      8000000000   
2                     NaN                                NaN             NaN   
3              -504000000                               5.12      1410000000   
4             11900000000                                4.6     21900000000   
5                     NaN                               3.33             NaN   

  Electricity - consumption(kWh) Electricity - production(kWh)      Exports  \
1                      652200000                     540000000    446000000   
2                            NaN                           NaN          NaN   
3                     6760000000                    5680000000    552400000   
4                    23610000000                   25760000000  32160000000   
5                      120900000                     130000000     30000000   

            GDP  ... Oil - proved reserves(bbl) Population  \
1   21500000000  ...                          0   29928987   
2           NaN  ...                        NaN        NaN   
3   17460000000  ...                  185500000    3563112   
4  212300000000  ...                11870000000   32531853   
5     500000000  ...                        NaN      57881   

  Public debt(% of GDP) Railways(km) Reserves of foreign exchange & gold  \
1                   NaN          NaN                                 NaN   
2                   NaN          NaN                                 NaN   
3                   NaN          447                          1206000000   
4                  37.4         3973                         43550000000   
5                   NaN          NaN                                 NaN   

  Telephones - main lines in use Telephones - mobile cellular  \
1                          33100                        15000   
2                            NaN                          NaN   
3                         255000                      1100000   
4                        2199600                      1447310   
5                          15000                         2377   

  Total fertility rate(children born/woman) Unemployment rate(%)  \
1                                      6.75                  NaN   
2                                       NaN                  NaN   
3                                      2.04                 14.8   
4                                      1.92                 25.4   
5                                      3.25                    6   

  Internet users per capita  
1                  0.000033  
2                       NaN  
3                  0.008420  
4                  0.015370  
5                       NaN  

[5 rows x 46 columns]

### From this point on, it would be ideal to do some visualization of the results but I'm gonna skip this and I'm gonna go ahead towards building our holiday destination recommendation model, ghax xbajt naqa issa brow

### The question we need to ask ourselves is, What economic and socioeconomic features should the ideal countries possess? Given our data, we can build a recommendation model by extracting the countries that correspond the best with these features:
##### - High GDP per capita
##### - Low unemployment rate
##### - High life expectancy at birth
##### - Low HIV/AIDS deaths
##### - Few people living with HIV/AIDS
##### - Low crime rate

### Other features one may include could be low debt, recent GDP growth or  recent industrial growth but these dont really fit well. Greece for example is a country with a lot of debt but its hard to not consider it a holiday destination when you see photos of Santorini. So we will need to keep an open mind.

In [285]:
recom_data = data[['Country','Death rate(deaths/1000 population)','HIV/AIDS - deaths','HIV/AIDS - people living with HIV/AIDS','HIV/AIDS - adult prevalence rate(%)','Life expectancy at birth(years)','GDP - per capita','Unemployment rate(%)','Internet users per capita']]

In [286]:
recom_data.head(30)

Country Death rate(deaths/1000 population)  \
1                   Afghanistan                              20.75   
2                      Akrotiri                                NaN   
3                       Albania                               5.12   
4                       Algeria                                4.6   
5                American Samoa                               3.33   
6                       Andorra                               6.07   
7                        Angola                               25.9   
8                      Anguilla                               5.43   
9                    Antarctica                                NaN   
10          Antigua and Barbuda                               5.44   
11                    Argentina                               7.56   
12                      Armenia                               8.16   
13                        Aruba                               6.57   
14  Ashmore and Cartier Islands                                NaN   
15                    Australia                               7.44   
16                      Austria                                9.7   
17                   Azerbaijan                               9.86   
18                  Bahamas The                               8.97   
19                      Bahrain                               4.08   
20                 Baker Island                                NaN   
21                   Bangladesh                                8.4   
22                     Barbados                               9.17   
23              Bassas da India                                NaN   
24                      Belarus                              14.15   
25                      Belgium                              10.22   
26                       Belize                               6.04   
27                        Benin                              13.76   
28                      Bermuda                               7.63   
29                       Bhutan                              12.94   
30                      Bolivia                               7.64   

   HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
1                NaN                                    NaN   
2                NaN                                    NaN   
3                NaN                                    NaN   
4                500                                   9100   
5                NaN                                    NaN   
6                NaN                                    NaN   
7              21000                                 240000   
8                NaN                                    NaN   
9                NaN                                    NaN   
10               NaN                                    NaN   
11              1500                                 130000   
12               200                                   2600   
13               NaN                                    NaN   
14               NaN                                    NaN   
15               NaN                                  14000   
16               100                                  10000   
17               100                                   1400   
18               200                                   5600   
19               200                                    600   
20               NaN                                    NaN   
21               650                                  13000   
22               200                                   2500   
23               NaN                                    NaN   
24              1000                                  15000   
25               100                                  10000   
26               NaN                                   3600   
27              5800                                  68000   
28               NaN                                    NaN   
29               NaN      

In [287]:
recom_data = recom_data.dropna(thresh=3) ## removed rows of data that have 3 or more NaN values

In [288]:
recom_data.head(30)

Country Death rate(deaths/1000 population)  \
1              Afghanistan                              20.75   
3                  Albania                               5.12   
4                  Algeria                                4.6   
5           American Samoa                               3.33   
6                  Andorra                               6.07   
7                   Angola                               25.9   
8                 Anguilla                               5.43   
10     Antigua and Barbuda                               5.44   
11               Argentina                               7.56   
12                 Armenia                               8.16   
13                   Aruba                               6.57   
15               Australia                               7.44   
16                 Austria                                9.7   
17              Azerbaijan                               9.86   
18             Bahamas The                               8.97   
19                 Bahrain                               4.08   
21              Bangladesh                                8.4   
22                Barbados                               9.17   
24                 Belarus                              14.15   
25                 Belgium                              10.22   
26                  Belize                               6.04   
27                   Benin                              13.76   
28                 Bermuda                               7.63   
29                  Bhutan                              12.94   
30                 Bolivia                               7.64   
31  Bosnia and Herzegovina                               8.44   
32                Botswana                              29.36   
34                  Brazil                               6.15   
36  British Virgin Islands                               4.42   
37                  Brunei                               3.42   

   HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
1                NaN                                    NaN   
3                NaN                                    NaN   
4                500                                   9100   
5                NaN                                    NaN   
6                NaN                                    NaN   
7              21000                                 240000   
8                NaN                                    NaN   
10               NaN                                    NaN   
11              1500                                 130000   
12               200                                   2600   
13               NaN                                    NaN   
15               NaN                                  14000   
16               100                                  10000   
17               100                                   1400   
18               200                                   5600   
19               200                                    600   
21               650                                  13000   
22               200                                   2500   
24              1000                                  15000   
25               100                                  10000   
26               NaN                                   3600   
27              5800                                  68000   
28               NaN                                    NaN   
29               NaN                                    100   
30               NaN                                   4900   
31               100                                    900   
32             33000                                 350000   
34             15000                                 660000   
36               NaN                                    NaN   
37               200                                    200   

   HIV/AIDS - adult prevalence rate(%) Life expec

In [289]:
recom_data = recom_data[pd.notnull(recom_data['Death rate(deaths/1000 population)'])] ## removing countries that have no inforation on death rate

In [290]:
recom_data.head(40)

Country Death rate(deaths/1000 population)  \
1                Afghanistan                              20.75   
3                    Albania                               5.12   
4                    Algeria                                4.6   
5             American Samoa                               3.33   
6                    Andorra                               6.07   
7                     Angola                               25.9   
8                   Anguilla                               5.43   
10       Antigua and Barbuda                               5.44   
11                 Argentina                               7.56   
12                   Armenia                               8.16   
13                     Aruba                               6.57   
15                 Australia                               7.44   
16                   Austria                                9.7   
17                Azerbaijan                               9.86   
18               Bahamas The                               8.97   
19                   Bahrain                               4.08   
21                Bangladesh                                8.4   
22                  Barbados                               9.17   
24                   Belarus                              14.15   
25                   Belgium                              10.22   
26                    Belize                               6.04   
27                     Benin                              13.76   
28                   Bermuda                               7.63   
29                    Bhutan                              12.94   
30                   Bolivia                               7.64   
31    Bosnia and Herzegovina                               8.44   
32                  Botswana                              29.36   
34                    Brazil                               6.15   
36    British Virgin Islands                               4.42   
37                    Brunei                               3.42   
38                  Bulgaria                              14.26   
39              Burkina Faso                              18.86   
40                     Burma                              12.15   
41                   Burundi                              17.43   
42                  Cambodia                               8.97   
43                  Cameroon                               15.4   
44                    Canada                               7.73   
45                Cape Verde                               6.62   
46            Cayman Islands                               4.81   
47  Central African Republic                              20.27   

   HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
1                NaN                                    NaN   
3                NaN                                    NaN   
4                500                                   9100   
5                NaN                                    NaN   
6                NaN                                    NaN   
7              21000                                 240000   
8                NaN                                    NaN   
10               NaN                                    NaN   
11              1500                                 130000   
12               200                                   2600   
13               NaN                                    NaN   
15               NaN                                  14000   
16               100                                  10000   
17               100                                   1400   
18               200                                   5600   
19               200                                    600   
21               650                                  13000   
22               200                                   2500   
24              1000                                  15000   
25           

### So when I think of poor countries, basically countries with a very low GDP per capita, I think of Mali for example. Given its such a poor country, there is a large probability that it is unsafe to travel. People there are probably living off scraps so foreigners might be preyed upon somehow. Another factor is sanitation, there is without a doubt a higher presence of diseases in these countries. So now lets classify the GDP per capita range that a 'poor' country falls upon (<= 1500) and we will print that list and see which countries pop up. We will analyze also other features of those countries and we will drop them from our list accordingly. L

In [291]:
recom_data['GDP - per capita'] = recom_data['GDP - per capita'].astype(int)
poor_c =  recom_data['GDP - per capita']<=1500
data5 = recom_data[poor_c]
data5.head(25) ### these countries will be dropped

Country Death rate(deaths/1000 population)  \
1                         Afghanistan                              20.75   
27                              Benin                              13.76   
29                             Bhutan                              12.94   
39                       Burkina Faso                              18.86   
41                            Burundi                              17.43   
45                         Cape Verde                               6.62   
47           Central African Republic                              20.27   
55                            Comoros                                8.4   
56   Congo Democratic Republic of the                              14.43   
57              Congo Republic of the                              14.82   
60                      Cote d'Ivoire                              14.94   
67                           Djibouti                              19.39   
70                         East Timor                                6.3   
75                            Eritrea                              13.53   
77                           Ethiopia                              15.06   
90                         Gaza Strip                               3.87   
104                     Guinea-Bissau                              16.53   
106                             Haiti                              12.34   
130                             Kenya                              14.65   
132                          Kiribati                               8.37   
133                       Korea North                               7.05   
141                           Liberia                              17.87   
148                        Madagascar                              11.35   
149                            Malawi                              23.39   
152                              Mali                              19.05   

    HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
1                 NaN                                    NaN   
27               5800                                  68000   
29                NaN                                    100   
39              29000                                 300000   
41              25000                                 250000   
45                225                                    775   
47              23000                                 260000   
55                NaN                                    NaN   
56             100000                                1100000   
57               9700                                  90000   
60              47000                                 570000   
67                690                                   9100   
70                NaN                                    NaN   
75               6300                                  60000   
77             120000                                1500000   
90                NaN                                    NaN   
104              1200                                  17000   
106             24000                                 280000   
130            150000                                1200000   
132               NaN                                    NaN   
133               NaN                                    NaN   
141              7200                                 100000   
148              7500                                 140000   
149             84000                                 900000   
152             12000                                 140000   

    HIV/AIDS - adult prevalence rate(%) Life expectancy at birth(years)  \
1                                  0.01                            42.9   
27                                  1.9                           50.51   
29                                  0.1                           54.39   
39                                  4.2                           43.92   
41           

In [292]:
recom_data = recom_data[recom_data['GDP - per capita'] > 1500]

In [293]:
recom_data.shape

(185, 9)

In [294]:
recom_data.head(50)

Country Death rate(deaths/1000 population)  \
3                  Albania                               5.12   
4                  Algeria                                4.6   
5           American Samoa                               3.33   
6                  Andorra                               6.07   
7                   Angola                               25.9   
8                 Anguilla                               5.43   
10     Antigua and Barbuda                               5.44   
11               Argentina                               7.56   
12                 Armenia                               8.16   
13                   Aruba                               6.57   
15               Australia                               7.44   
16                 Austria                                9.7   
17              Azerbaijan                               9.86   
18             Bahamas The                               8.97   
19                 Bahrain                               4.08   
21              Bangladesh                                8.4   
22                Barbados                               9.17   
24                 Belarus                              14.15   
25                 Belgium                              10.22   
26                  Belize                               6.04   
28                 Bermuda                               7.63   
30                 Bolivia                               7.64   
31  Bosnia and Herzegovina                               8.44   
32                Botswana                              29.36   
34                  Brazil                               6.15   
36  British Virgin Islands                               4.42   
37                  Brunei                               3.42   
38                Bulgaria                              14.26   
40                   Burma                              12.15   
42                Cambodia                               8.97   
43                Cameroon                               15.4   
44                  Canada                               7.73   
46          Cayman Islands                               4.81   
48                    Chad                              16.41   
49                   Chile                               5.76   
50                   China                               6.94   
54                Colombia                               5.59   
59              Costa Rica                               4.33   
61                 Croatia                              11.38   
62                    Cuba                               7.19   
63                  Cyprus                               7.64   
64          Czech Republic                              10.54   
65                 Denmark                              10.43   
68                Dominica                               6.81   
69      Dominican Republic                               7.35   
71                 Ecuador                               4.24   
72                   Egypt                               5.26   
73             El Salvador                               5.85   
74       Equatorial Guinea                                 12   
76                 Estonia                              13.21   

   HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
3                NaN                                    NaN   
4                500                                   9100   
5                NaN                                    NaN   
6                NaN                                    NaN   
7              21000                                 240000   
8                NaN                                    NaN   
10               NaN                                    NaN   
11              1500                                 130000   
12               200                                   2600   
13               NaN                                    NaN   
15        

### Next we will filter out countries who have a life expectancy less than that of 55 years

In [295]:
recom_data['Life expectancy at birth(years)'] = recom_data['Life expectancy at birth(years)'].astype(float)
recom_data = recom_data[recom_data['Life expectancy at birth(years)'] >= 55]
recom_data.shape

(173, 9)

### Next we will filter out countries with more than a 15% unemployment rate

In [296]:
recom_data['Unemployment rate(%)'] = recom_data['Unemployment rate(%)'].astype(float)
recom_data = recom_data[recom_data['Unemployment rate(%)'] < 15]
recom_data.shape

(120, 9)

In [297]:
recom_data.head(20)

Country Death rate(deaths/1000 population)  \
3                  Albania                               5.12   
5           American Samoa                               3.33   
6                  Andorra                               6.07   
8                 Anguilla                               5.43   
10     Antigua and Barbuda                               5.44   
11               Argentina                               7.56   
13                   Aruba                               6.57   
15               Australia                               7.44   
16                 Austria                                9.7   
17              Azerbaijan                               9.86   
18             Bahamas The                               8.97   
22                Barbados                               9.17   
24                 Belarus                              14.15   
25                 Belgium                              10.22   
26                  Belize                               6.04   
28                 Bermuda                               7.63   
30                 Bolivia                               7.64   
34                  Brazil                               6.15   
36  British Virgin Islands                               4.42   
37                  Brunei                               3.42   

   HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
3                NaN                                    NaN   
5                NaN                                    NaN   
6                NaN                                    NaN   
8                NaN                                    NaN   
10               NaN                                    NaN   
11              1500                                 130000   
13               NaN                                    NaN   
15               NaN                                  14000   
16               100                                  10000   
17               100                                   1400   
18               200                                   5600   
22               200                                   2500   
24              1000                                  15000   
25               100                                  10000   
26               NaN                                   3600   
28               NaN                                    NaN   
30               NaN                                   4900   
34             15000                                 660000   
36               NaN                                    NaN   
37               200                                    200   

   HIV/AIDS - adult prevalence rate(%)  Life expectancy at birth(years)  \
3                                  NaN                            77.24   
5                                  NaN                            75.84   
6                                  NaN                            83.51   
8                                  NaN                            77.11   
10                                 NaN                            71.90   
11                                 0.7                            75.91   
13                                 NaN                            79.14   
15                                 0.1                            80.39   
16                                 0.3                            78.92   
17                                 0.1                            63.35   
18                                   3                            65.54   
22                                 1.5                            71.41   
24                                 0.3                            68.72   
25                                 0.2                            78.62   
26                                 2.4                            67.49   
28                                 NaN                            77.79   
30                                 0.1                

### Now we would like to view the countries that have a NaN value for 'HIV adult prevalence rate'. Some of these countries, albeit having a NaN values are quite decent and popular holiday destionations and shouldnt be removed from our recommendation list. So lets seperate them

In [298]:
recom_data['HIV/AIDS - adult prevalence rate(%)'] = recom_data['HIV/AIDS - adult prevalence rate(%)'].astype(float)
null_data = recom_data[recom_data['HIV/AIDS - adult prevalence rate(%)'].isnull()]
null_data.shape



(31, 9)

In [299]:
null_data.head(40)

Country Death rate(deaths/1000 population)  \
3                      Albania                               5.12   
5               American Samoa                               3.33   
6                      Andorra                               6.07   
8                     Anguilla                               5.43   
10         Antigua and Barbuda                               5.44   
13                       Aruba                               6.57   
28                     Bermuda                               7.63   
36      British Virgin Islands                               4.42   
46              Cayman Islands                               4.81   
79             #European Union                               10.1   
81               Faroe Islands                               8.69   
86            French Polynesia                               4.63   
94                   Gibraltar                               9.18   
97                   Greenland                               7.77   
98                     Grenada                               7.17   
102                   Guernsey                               9.95   
125                     Jersey                               9.19   
143              Liechtenstein                               7.06   
146                      Macau                               4.23   
154                Man Isle of                              11.26   
166                 Montserrat                               7.17   
186                      Palau                               6.85   
196                Puerto Rico                               7.86   
202               Saint Helena                               6.43   
203      Saint Kitts and Nevis                               8.47   
205  Saint Pierre and Miquelon                                6.7   
208                 San Marino                               8.07   
232                     Taiwan                               6.38   
238                      Tonga                               5.35   
244   Turks and Caicos Islands                               4.28   
256             Virgin Islands                               6.26   

    HIV/AIDS - deaths HIV/AIDS - people living with HIV/AIDS  \
3                 NaN                                    NaN   
5                 NaN                                    NaN   
6                 NaN                                    NaN   
8                 NaN                                    NaN   
10                NaN                                    NaN   
13                NaN                                    NaN   
28                NaN                                    NaN   
36                NaN                                    NaN   
46                NaN                                    NaN   
79                NaN                                    NaN   
81                NaN                                    NaN   
86                NaN                                    NaN   
94                NaN                                    NaN   
97                NaN                                    100   
98                NaN                                    NaN   
102               NaN                                    NaN   
125               NaN                                    NaN   
143               NaN                                    NaN   
146               NaN                                    NaN   
154               NaN                                    NaN   
166               NaN                                    NaN   
186               NaN                                    NaN   
196               NaN                                   7397   
202               NaN                                    NaN   
203               NaN                                    NaN   
205               NaN                                    NaN   
208               NaN                                    NaN   
232               NaN 

### Next we will filter out countries with an HIV prevalence rate of >=2%

In [302]:
recom_data['HIV/AIDS - adult prevalence rate(%)'] = recom_data['HIV/AIDS - adult prevalence rate(%)'].astype(float)
k_data = recom_data[(recom_data['HIV/AIDS - adult prevalence rate(%)'] < 2) ] 
k_data.shape

(85, 9)

In [303]:
final_data = pd.concat([k_data, null_data])  ## MERGING THE 2 DATAFRAMES TOGETHER

final_data.shape

(116, 9)

In [306]:
final_data.head(50)

Country Death rate(deaths/1000 population) HIV/AIDS - deaths  \
11        Argentina                               7.56              1500   
15        Australia                               7.44               NaN   
16          Austria                                9.7               100   
17       Azerbaijan                               9.86               100   
22         Barbados                               9.17               200   
24          Belarus                              14.15              1000   
25          Belgium                              10.22               100   
30          Bolivia                               7.64               NaN   
34           Brazil                               6.15             15000   
37           Brunei                               3.42               200   
38         Bulgaria                              14.26               100   
40            Burma                              12.15             20000   
44           Canada                               7.73              1500   
49            Chile                               5.76              1400   
50            China                               6.94             44000   
54         Colombia                               5.59              3600   
59       Costa Rica                               4.33               900   
61          Croatia                              11.38                10   
62             Cuba                               7.19               200   
63           Cyprus                               7.64               NaN   
64   Czech Republic                              10.54                10   
65          Denmark                              10.43               100   
71          Ecuador                               4.24              1700   
72            Egypt                               5.26               700   
73      El Salvador                               5.85              2200   
76          Estonia                              13.21               200   
82             Fiji                               5.65               200   
83          Finland                               9.79               100   
84           France                               9.08              1000   
92          Germany                              10.55              1000   
96           Greece                              10.15               100   
101       Guatemala                               6.81              5800   
110       Hong Kong                               5.98               200   
112         Hungary                              13.19               100   
113         Iceland                               6.68               100   
114           India                               8.28            310000   
115       Indonesia                               6.25              2400   
116            Iran                               5.55               800   
118         Ireland                               7.85               100   
119          Israel                               6.18               100   
120           Italy                               10.3              1000   
123           Japan                               8.95               500   
129      Kazakhstan                               9.46               200   
134     Korea South                               6.26               200   
135          Kuwait                               2.42               NaN   
137            Laos                              11.83               200   
138          Latvia                               13.7               500   
144       Lithuania                              10.92               200   
145      Luxembourg                               8.41               100   
150        Malaysia                               5.06              2000   

    HIV/AIDS - people living with HIV/AIDS  \
11                                  130000   
15                                   14000 

### We can also filter out countries with very little internet usage per capita but this doesnt necessarily mean that they are bad holiday destionations. Ideally we would have more data but given the data that we have, we can with some certainty assume that the 116 countries left are all good holiday options. Some better than others, depending on your tastes and budget. Another thing we could do is we can rank the countries. We can insert a rank number instead of the value, then we can add the total of the rank values to give out a score. The highest score is the most recommended country out of our recommended countries and so and so for lowest score. We can also fill some of the missing values by doing seperate research so we can obtain a complete dataset.